channeldb/messagestore.go

```diff
@@ -0,0 +1,227 @@
+		if err != nil {
+			return err
+		}
+
+		var b bytes.Buffer
+		if err := msg.Encode(&b, 0); err != nil {
+			return err
+		}
+
+		// Generate next sequence number to preserver the message order.
+		sequence, err := peerBucket.NextSequence()
+		if err != nil {
+			return err
+		}
+
+		sequenceKey := make([]byte, 8)
+		binary.BigEndian.PutUint64(sequenceKey, sequence)
+		return messageBucket.Put(sequenceKey, b.Bytes())
+	}); err != nil {
+		return err
+	}
+
+	// now message with this code exist in the storage, sync cache.
+	s.cache[msg.Command()] = true
+	return nil
+}
+
+// Remove removes the message buckets by message type/code. Such specific
+// deletion is due to the fact of logic of retransmission subsystem, where we
+// need remove messages not one by one but in contrary the whole group of
+// messages.
+func (s *MessagesStore) Remove(codes ...lnwire.Code) error {
+	storedCodes := make([]lnwire.Code, 0, len(codes))
+
+	// Reduce number of boltdb calls by using cache to filter the codes
+	// that should be removed.
+	for _, code := range codes {
+		if _, ok := s.cache[code]; ok {
+			storedCodes = append(storedCodes, code)
+		}
+	}
+
+	if err := s.db.Batch(func(tx *bolt.Tx) error {
+		peerBucketKey := s.getPeerBucketKey()
+		peerBucket, err := tx.CreateBucketIfNotExists(peerBucketKey)
+		if err != nil {
+			return err
+		}
+
+		for _, code := range storedCodes {
+			key := make([]byte, 4)
+			binary.BigEndian.PutUint32(key, uint32(code))
+
+			err := peerBucket.DeleteBucket(key)
+			if err != nil && bolt.ErrBucketNotFound != err {
+				return err
+			}
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	// now message with this code are not exist in the storage, sync cache.
+	for _, code := range storedCodes {
+		delete(s.cache, code)
+	}
+
+	return nil
+}
+
+// Get retrieves messages from storage and sort them in the order in which
+// they were originally added, sorting is needed for retransmission
+// subsystem, as far this messages should be resent to remote peer in the same
+// order as they were sent originally.
+func (s *MessagesStore) Get() ([]lnwire.Message, error) {
+	var indexes []int
+	messageMap := make(map[uint64]lnwire.Message)
+
+	if err := s.db.View(func(tx *bolt.Tx) error {
+		peerBucketKey := s.getPeerBucketKey()
+		peerBucket := tx.Bucket(peerBucketKey)
+		if peerBucket == nil {
+			return nil
+		}
+
+		// Iterate over messages buckets.
+		return peerBucket.ForEach(func(k, v []byte) error {
+			// Skip non buckets fields.
+			if v != nil {
+				return nil
+			}
+
+			messageBucket := peerBucket.Bucket(k)
+			if messageBucket == nil {
+				return ErrMessageNotFound
+			}
+
+			// Retrieve code/type of messages in the bucket.
+			code := lnwire.Code(binary.BigEndian.Uint32(k))
+
+			return messageBucket.ForEach(func(k, v []byte) error {
+				// Skip non messages.
+				if v == nil {
+					return nil
+				}
+
+				// Add indexes in additional array, because
```

Comment:
See the comment above about re-working the schema to eliminate this sorting step.